iceberg/src/main/scala/org/apache/spark/sql/delta/icebergShaded/HudiConverter.scala

```diff
@@ -0,0 +1,363 @@
+                            "stackTrace" -> ExceptionUtils.getStackTrace(e)
+                          )
+                        )
+                    }
+                    currentConversion.set(null)
+                    // Pick next snapshot to convert if there's a new one
+                    snapshotAndTxn = getNextSnapshot
+                  }
+              } finally {
+                // shuttingdown thread
+                asyncThreadLock.synchronized {
+                  asyncConverterThreadActive = false
+                }
+              }
+
+          // Get a snapshot to convert from the hudiQueue. Sets the queue to null after.
+          private def getNextSnapshot: (Snapshot, OptimisticTransactionImpl) =
+            asyncThreadLock.synchronized {
+              val potentialSnapshotAndTxn = standbyConversion.get()
+              currentConversion.set(potentialSnapshotAndTxn)
+              standbyConversion.compareAndSet(potentialSnapshotAndTxn, null)
+              if (potentialSnapshotAndTxn == null) {
+                asyncConverterThreadActive = false
+              }
+              potentialSnapshotAndTxn
+            }
+        }
+        asyncConverterThread.start()
+        asyncConverterThreadActive = true
+      }
+    }
+
+    // If there already was a snapshot waiting to be converted, log that snapshot info.
+    if (previouslyQueued != null) {
+      recordDeltaEvent(
+        snapshotToConvert.deltaLog,
+        "delta.hudi.conversion.async.backlog",
+        data = Map(
+          "newVersion" -> snapshotToConvert.version,
+          "replacedVersion" -> previouslyQueued._1.version)
+      )
+    }
+  }
+
+  /**
+   * Convert the specified snapshot into Hudi for the given catalogTable
+   * @param snapshotToConvert the snapshot that needs to be converted to Hudi
+   * @param catalogTable the catalogTable this conversion targets.
+   * @return Converted Delta version and commit timestamp
+   */
+  override def convertSnapshot(
+      snapshotToConvert: Snapshot, catalogTable: CatalogTable): Option[(Long, Long)] = {
+    if (!UniversalFormat.hudiEnabled(snapshotToConvert.metadata)) {
+      return None
+    }
+    convertSnapshot(snapshotToConvert, None, catalogTable)
+  }
+
+  /**
+   * Convert the specified snapshot into Hudi when performing an OptimisticTransaction
+   * on a delta table.
+   * @param snapshotToConvert the snapshot that needs to be converted to Hudi
+   * @param txn               the transaction that triggers the conversion. It must
+   *                          contain the catalogTable this conversion targets.
+   * @return Converted Delta version and commit timestamp
+   */
+  override def convertSnapshot(
+      snapshotToConvert: Snapshot, txn: OptimisticTransactionImpl): Option[(Long, Long)] = {
+    if (!UniversalFormat.hudiEnabled(snapshotToConvert.metadata)) {
+      return None
+    }
+    txn.catalogTable match {
```

Comment:
Iceberg almost always require a catalog so we had this requirement. If thats not true for Hudi, I would suggest to release this contraint, otherwise it wont work if we access/write to the Delta table by path. 

However on the other hand I get the fact that we need to pass in the tableName to Hudi table creator. 

Can we add a TODO here saying requiring a catalogTable is a constraint that we can release in the future ?

Reply:
We don't require a catalog for Hudi. I just figured if there is a name present we can use it when creating the table.